File: agents/improvement_cycle.py
"""
DEPRECATED: This Python agent is being migrated to Go.

Per the project's language policy, all agents must be implemented in:
- C++, DLLs, GO, Rust, or R

Python is no longer permitted for agent implementations.

Migration Status:
- This file will be replaced by improvement_cycle.go
- See agents/LANGUAGE_POLICY.md for details

Original: Improvement Cycle Orchestrator
"""

import asyncio
import json
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any, Dict, List, Optional
from enum import Enum

import structlog
from pydantic import BaseModel, Field

from .base_agent import BaseAgent, AgentConfig, TaskResult
from .llm.gemini_service import GeminiService, GeminiConfig

logger = structlog.get_logger()


class CyclePhase(Enum):
    """Phases of the improvement cycle."""
    ANALYSIS = "analysis"
    PLANNING = "planning"
    IMPLEMENTATION = "implementation"
    VALIDATION = "validation"
    DEPLOYMENT = "deployment"
    COMPLETED = "completed"
    FAILED = "failed"


class ImprovementStatus(BaseModel):
    """Status of an improvement cycle."""
    cycle_id: str
    phase: CyclePhase
    started_at: datetime
    completed_at: Optional[datetime] = None
    patterns_identified: int = 0
    improvements_planned: int = 0
    improvements_implemented: int = 0
    improvements_validated: int = 0
    success_rate: float = 0.0
    errors: List[str] = Field(default_factory=list)


class ImprovementPlan(BaseModel):
    """Plan for implementing improvements."""
    plan_id: str
    patterns: List[Dict[str, Any]]
    improvements: List[Dict[str, Any]]
    priority_order: List[str]
    estimated_effort: Dict[str, float]
    risks: List[Dict[str, Any]]
    success_criteria: Dict[str, Any]
    created_at: datetime = Field(default_factory=datetime.now)


class ImprovementCycle:
    """
    Orchestrates the auto-improvement cycle.
    
    This class coordinates multiple specialized agents to continuously
    analyze, plan, implement, and validate improvements to the system.
    """
    
    def __init__(
        self,
        config_path: Optional[str] = None,
        llm_service: Optional[GeminiService] = None
    ):
        self.config = self._load_config(config_path)
        self.llm_service = llm_service or GeminiService()
        self.agents: Dict[str, BaseAgent] = {}
        self.cycle_history: List[ImprovementStatus] = []
        self.current_cycle: Optional[ImprovementStatus] = None
        self._running = False
        
        # Phase success criteria
        self._phase_criteria = {
            CyclePhase.ANALYSIS: {
                "min_patterns": 3,
                "confidence_threshold": 0.8,
            },
            CyclePhase.PLANNING: {
                "plan_completeness": 0.9,
                "risk_assessment": True,
            },
            CyclePhase.IMPLEMENTATION: {
                "code_quality": 0.85,
                "test_coverage": 0.8,
            },
            CyclePhase.VALIDATION: {
                "all_tests_pass": True,
                "performance_improvement": 0.1,
            },
            CyclePhase.DEPLOYMENT: {
                "deployment_success": True,
                "rollback_plan": True,
            },
        }
        
        logger.info("Improvement cycle orchestrator created")
    
    def _load_config(self, config_path: Optional[str]) -> Dict[str, Any]:
        """Load configuration from file."""
        if config_path:
            path = Path(config_path)
            if path.exists():
                return json.loads(path.read_text(encoding="utf-8"))
        
        # Default config path
        default_path = Path("agents/config.json")
        if default_path.exists():
            return json.loads(default_path.read_text(encoding="utf-8"))
        
        # Return minimal config
        return {
            "gemini": {
                "model": "gemini-3-flash-preview",
                "rate_limits": {"rpm": 15, "tpm": 1000000, "rpd": 1500}
            }
        }
    
    async def initialize(self) -> None:
        """Initialize the improvement cycle."""
        logger.info("Initializing improvement cycle")
        
        # Initialize LLM service
        await self.llm_service.initialize()
        
        # Initialize agents (will be implemented in separate modules)
        # await self._initialize_agents()
        
        logger.info("Improvement cycle initialized")
    
    async def _initialize_agents(self) -> None:
        """Initialize all specialized agents."""
        from .research_agent import ResearchAgent
        from .analysis_agent import AnalysisAgent
        from .code_agent import CodeAgent
        from .test_agent import TestAgent
        
        agent_configs = self.config.get("agents", {})
        
        # Create agents
        agents_to_create = [
            ("research", ResearchAgent),
            ("analysis", AnalysisAgent),
            ("code", CodeAgent),
            ("test", TestAgent),
        ]
        
        for agent_type, agent_class in agents_to_create:
            config_data = agent_configs.get(f"{agent_type}_agent", {})
            config = AgentConfig(
                name=config_data.get("name", f"{agent_type.title()}Agent"),
                description=config_data.get("description", ""),
                system_prompt=config_data.get("system_prompt", ""),
                tools=config_data.get("tools", []),
                max_iterations=config_data.get("max_iterations", 5),
            )
            
            agent = agent_class(config, self.llm_service)
            await agent.initialize()
            self.agents[agent_type] = agent
            
            logger.info(f"Agent initialized", agent_type=agent_type)
    
    async def run_cycle(self) -> ImprovementStatus:
        """
        Run a complete improvement cycle.
        
        Returns:
            ImprovementStatus with cycle results
        """
        if self._running:
            raise RuntimeError("Improvement cycle is already running")
        
        self._running = True
        cycle_id = f"cycle_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        
        self.current_cycle = ImprovementStatus(
            cycle_id=cycle_id,
            phase=CyclePhase.ANALYSIS,
            started_at=datetime.now()
        )
        
        logger.info("Starting improvement cycle", cycle_id=cycle_id)
        
        try:
            # Phase 1: Analysis
            await self._run_analysis_phase()
            
            # Phase 2: Planning
            await self._run_planning_phase()
            
            # Phase 3: Implementation
            await self._run_implementation_phase()
            
            # Phase 4: Validation
            await self._run_validation_phase()
            
            # Phase 5: Deployment
            await self._run_deployment_phase()
            
            # Mark as completed
            self.current_cycle.phase = CyclePhase.COMPLETED
            self.current_cycle.completed_at = datetime.now()
            
            # Calculate success rate
            if self.current_cycle.improvements_planned > 0:
                self.current_cycle.success_rate = (
                    self.current_cycle.improvements_validated /
                    self.current_cycle.improvements_planned
                )
            
            logger.info("Improvement cycle completed",
                       cycle_id=cycle_id,
                       success_rate=self.current_cycle.success_rate)
            
        except Exception as e:
            self.current_cycle.phase = CyclePhase.FAILED
            self.current_cycle.completed_at = datetime.now()
            self.current_cycle.errors.append(str(e))
            
            logger.error("Improvement cycle failed",
                        cycle_id=cycle_id,
                        error=str(e))
        
        finally:
            self._running = False
            self.cycle_history.append(self.current_cycle)
            
            # Save cycle history
            await self._save_cycle_history()
        
        return self.current_cycle
    
    async def _run_analysis_phase(self) -> None:
        """Run the analysis phase."""
        logger.info("Running analysis phase")
        self.current_cycle.phase = CyclePhase.ANALYSIS
        
        # Analyze current performance
        performance_data = await self._collect_performance_data()
        
        # Identify patterns using Gemini
        patterns = await self._identify_patterns(performance_data)
        
        self.current_cycle.patterns_identified = len(patterns)
        
        # Check success criteria
        if len(patterns) < self._phase_criteria[CyclePhase.ANALYSIS]["min_patterns"]:
            raise ValueError(f"Insufficient patterns identified: {len(patterns)}")
        
        logger.info("Analysis phase completed", patterns=len(patterns))
    
    async def _run_planning_phase(self) -> None:
        """Run the planning phase."""
        logger.info("Running planning phase")
        self.current_cycle.phase = CyclePhase.PLANNING
        
        # Generate improvement plan
        plan = await self._generate_improvement_plan()
        
        self.current_cycle.improvements_planned = len(plan.improvements)
        
        logger.info("Planning phase completed", improvements=len(plan.improvements))
    
    async def _run_implementation_phase(self) -> None:
        """Run the implementation phase."""
        logger.info("Running implementation phase")
        self.current_cycle.phase = CyclePhase.IMPLEMENTATION
        
        # Implement improvements
        implemented = await self._implement_improvements()
        
        self.current_cycle.improvements_implemented = implemented
        
        logger.info("Implementation phase completed", implemented=implemented)
    
    async def _run_validation_phase(self) -> None:
        """Run the validation phase."""
        logger.info("Running validation phase")
        self.current_cycle.phase = CyclePhase.VALIDATION
        
        # Validate improvements
        validated = await self._validate_improvements()
        
        self.current_cycle.improvements_validated = validated
        
        logger.info("Validation phase completed", validated=validated)
    
    async def _run_deployment_phase(self) -> None:
        """Run the deployment phase."""
        logger.info("Running deployment phase")
        self.current_cycle.phase = CyclePhase.DEPLOYMENT
        
        # Deploy validated improvements
        await self._deploy_improvements()
        
        logger.info("Deployment phase completed")
    
    async def _collect_performance_data(self) -> Dict[str, Any]:
        """Collect performance data from all agents."""
        data = {
            "timestamp": datetime.now().isoformat(),
            "agents": {},
            "system_metrics": {},
        }
        
        # Collect from each agent
        for agent_type, agent in self.agents.items():
            try:
                performance = agent.get_performance_summary()
                data["agents"][agent_type] = performance
            except Exception as e:
                logger.warning(f"Failed to collect from {agent_type}", error=str(e))
        
        # Add system metrics
        data["system_metrics"] = {
            "llm_rate_limits": self.llm_service.get_rate_limit_status(),
            "cycle_history_length": len(self.cycle_history),
        }
        
        return data
    
    async def _identify_patterns(self, performance_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Identify patterns in performance data using Gemini."""
        prompt = """
Analyze the following performance data from an AI agent system.
Identify patterns, bottlenecks, and opportunities for improvement.

Look for:
1. Recurring error patterns
2. Performance bottlenecks
3. Resource usage inefficiencies
4. Opportunities for optimization

Provide your analysis in JSON format.
"""
        
        analysis = await self.llm_service.analyze_with_context(
            analysis_prompt=prompt,
            context=performance_data
        )
        
        # Extract patterns from analysis
        patterns = []
        if "patterns" in analysis:
            for pattern in analysis["patterns"]:
                patterns.append({
                    "pattern": pattern,
                    "confidence": analysis.get("confidence", 0.5),
                    "source": "gemini_analysis"
                })
        
        # Store patterns in memory
        for pattern in patterns:
            for agent in self.agents.values():
                agent.memory.store_pattern(pattern)
        
        return patterns
    
    async def _generate_improvement_plan(self) -> ImprovementPlan:
        """Generate improvement plan based on identified patterns."""
        # Collect recent patterns from all agents
        all_patterns = []
        for agent in self.agents.values():
            all_patterns.extend(agent.memory.get_recent_patterns(limit=20))
        
        prompt = f"""
Based on the following patterns identified in an AI agent system,
create an improvement plan with specific actions.

Patterns:
{json.dumps(all_patterns, indent=2)}

Create a plan with:
1. Specific improvements for each pattern
2. Priority ordering (high/medium/low)
3. Estimated effort (hours)
4. Risk assessment
5. Success criteria

Provide your plan in JSON format.
"""
        
        plan_data = await self.llm_service.analyze_with_context(
            analysis_prompt=prompt,
            context={"patterns": all_patterns}
        )
        
        plan = ImprovementPlan(
            plan_id=f"plan_{datetime.now().strftime('%Y%m%d_%H%M%S')}",
            patterns=all_patterns,
            improvements=plan_data.get("improvements", []),
            priority_order=plan_data.get("priority_order", []),
            estimated_effort=plan_data.get("estimated_effort", {}),
            risks=plan_data.get("risks", []),
            success_criteria=plan_data.get("success_criteria", {}),
        )
        
        return plan
    
    async def _implement_improvements(self) -> int:
        """Implement improvements from the plan."""
        # This would coordinate with the code agent
        # For now, return a placeholder
        return 0
    
    async def _validate_improvements(self) -> int:
        """Validate implemented improvements."""
        # This would coordinate with the test agent
        # For now, return a placeholder
        return 0
    
    async def _deploy_improvements(self) -> None:
        """Deploy validated improvements."""
        # This would handle deployment logic
        pass
    
    async def _save_cycle_history(self) -> None:
        """Save cycle history to disk."""
        history_path = Path("agents/memory/cycle_history.json")
        history_path.parent.mkdir(parents=True, exist_ok=True)
        
        history_data = [
            cycle.model_dump(mode="json")
            for cycle in self.cycle_history
        ]
        
        history_path.write_text(
            json.dumps(history_data, indent=2, default=str),
            encoding="utf-8"
        )
        
        logger.info("Cycle history saved", path=str(history_path))
    
    def get_status(self) -> Dict[str, Any]:
        """Get current improvement cycle status."""
        return {
            "running": self._running,
            "current_cycle": (
                self.current_cycle.model_dump(mode="json")
                if self.current_cycle else None
            ),
            "total_cycles": len(self.cycle_history),
            "successful_cycles": sum(
                1 for c in self.cycle_history
                if c.phase == CyclePhase.COMPLETED
            ),
            "agents_initialized": len(self.agents),
            "llm_rate_limits": self.llm_service.get_rate_limit_status(),
        }
    
    async def shutdown(self) -> None:
        """Shutdown the improvement cycle."""
        logger.info("Shutting down improvement cycle")
        
        # Shutdown all agents
        for agent_type, agent in self.agents.items():
            try:
                await agent.shutdown()
            except Exception as e:
                logger.warning(f"Failed to shutdown {agent_type}", error=str(e))
        
        # Shutdown LLM service
        await self.llm_service.shutdown()
        
        self._running = False
        logger.info("Improvement cycle shut down")

File: agents/pkg/improvement_cycle.go
package pkg

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// CyclePhase represents phases of the improvement cycle
type CyclePhase string

const (
	CyclePhaseAnalysis     CyclePhase = "analysis"
	CyclePhasePlanning     CyclePhase = "planning"
	CyclePhaseImplementation CyclePhase = "implementation"
	CyclePhaseValidation   CyclePhase = "validation"
	CyclePhaseDeployment   CyclePhase = "deployment"
	CyclePhaseCompleted    CyclePhase = "completed"
	CyclePhaseFailed       CyclePhase = "failed"
)

// ImprovementStatus contains status of an improvement cycle
type ImprovementStatus struct {
	CycleID                string                 `json:"cycle_id"`
	Phase                  CyclePhase             `json:"phase"`
	StartedAt              time.Time              `json:"started_at"`
	CompletedAt            time.Time              `json:"completed_at,omitempty"`
	PatternsIdentified     int                    `json:"patterns_identified"`
	ImprovementsPlanned    int                    `json:"improvements_planned"`
	ImprovementsImplemented int                   `json:"improvements_implemented"`
	ImprovementsValidated  int                    `json:"improvements_validated"`
	SuccessRate            float64                `json:"success_rate"`
	Errors                 []string               `json:"errors"`
}

// ImprovementPlan contains plan for implementing improvements
type ImprovementPlan struct {
	PlanID         string                   `json:"plan_id"`
	Patterns       []map[string]interface{} `json:"patterns"`
	Improvements   []map[string]interface{} `json:"improvements"`
	PriorityOrder  []string                 `json:"priority_order"`
	EstimatedEffort map[string]float64      `json:"estimated_effort"`
	Risks          []map[string]interface{} `json:"risks"`
	SuccessCriteria map[string]interface{}  `json:"success_criteria"`
	CreatedAt      time.Time                `json:"created_at"`
}

// PhaseCriteria defines success criteria for each cycle phase
type PhaseCriteria struct {
	MinPatterns           int     `json:"min_patterns"`
	ConfidenceThreshold   float64 `json:"confidence_threshold"`
	PlanCompleteness      float64 `json:"plan_completeness"`
	RiskAssessment        bool    `json:"risk_assessment"`
	CodeQuality           float64 `json:"code_quality"`
	TestCoverage          float64 `json:"test_coverage"`
	AllTestsPass          bool    `json:"all_tests_pass"`
	PerformanceImprovement float64 `json:"performance_improvement"`
	DeploymentSuccess     bool    `json:"deployment_success"`
	RollbackPlan          bool    `json:"rollback_plan"`
}

// ImprovementCycle orchestrates the auto-improvement cycle
type ImprovementCycle struct {
	Config        map[string]interface{}
	Agents        map[string]Agent
	CycleHistory  []ImprovementStatus
	CurrentCycle  *ImprovementStatus
	mu            sync.Mutex
	running       bool

	phaseCriteria map[CyclePhase]PhaseCriteria
}

// NewImprovementCycle creates a new ImprovementCycle
func NewImprovementCycle(configPath string) (*ImprovementCycle, error) {
	cycle := &ImprovementCycle{
		Agents:       make(map[string]Agent),
		CycleHistory: make([]ImprovementStatus, 0),
		phaseCriteria: map[CyclePhase]PhaseCriteria{
			CyclePhaseAnalysis: {
				MinPatterns:         3,
				ConfidenceThreshold: 0.8,
			},
			CyclePhasePlanning: {
				PlanCompleteness: 0.9,
				RiskAssessment:   true,
			},
			CyclePhaseImplementation: {
				CodeQuality:  0.85,
				TestCoverage: 0.8,
			},
			CyclePhaseValidation: {
				AllTestsPass:          true,
				PerformanceImprovement: 0.1,
			},
			CyclePhaseDeployment: {
				DeploymentSuccess: true,
				RollbackPlan:      true,
			},
		},
	}

	// Load configuration
	if err := cycle.loadConfig(configPath); err != nil {
		return nil, err
	}

	return cycle, nil
}

// loadConfig loads configuration from file
func (c *ImprovementCycle) loadConfig(configPath string) error {
	// Try provided path first
	if configPath != "" {
		data, err := os.ReadFile(configPath)
		if err == nil {
			return json.Unmarshal(data, &c.Config)
		}
	}

	// Try default path
	defaultPath := filepath.Join("agents", "config.json")
	data, err := os.ReadFile(defaultPath)
	if err == nil {
		return json.Unmarshal(data, &c.Config)
	}

	// Use minimal default config
	c.Config = map[string]interface{}{
		"gemini": map[string]interface{}{
			"model": "gemini-3-flash-preview",
			"rate_limits": map[string]interface{}{
				"rpm": 15,
				"tpm": 1000000,
				"rpd": 1500,
			},
		},
	}

	return nil
}

// Initialize initializes the improvement cycle
func (c *ImprovementCycle) Initialize() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Initialize LLM service and agents will be implemented in separate modules

	return nil
}

// RunCycle runs a complete improvement cycle
func (c *ImprovementCycle) RunCycle() (*ImprovementStatus, error) {
	c.mu.Lock()
	if c.running {
		c.mu.Unlock()
		return nil, fmt.Errorf("improvement cycle is already running")
	}
	c.running = true
	c.mu.Unlock()

	defer func() {
		c.mu.Lock()
		c.running = false
		c.mu.Unlock()
	}()

	cycleID := fmt.Sprintf("cycle_%s", time.Now().Format("20060102_150405"))

	c.CurrentCycle = &ImprovementStatus{
		CycleID:   cycleID,
		Phase:     CyclePhaseAnalysis,
		StartedAt: time.Now(),
		Errors:    make([]string, 0),
	}

	var err error

	// Phase 1: Analysis
	if err = c.runAnalysisPhase(); err != nil {
		c.CurrentCycle.Phase = CyclePhaseFailed
		c.CurrentCycle.CompletedAt = time.Now()
		c.CurrentCycle.Errors = append(c.CurrentCycle.Errors, err.Error())
		c.CycleHistory = append(c.CycleHistory, *c.CurrentCycle)
		_ = c.saveCycleHistory()
		return c.CurrentCycle, err
	}

	// Phase 2: Planning
	if err = c.runPlanningPhase(); err != nil {
		c.CurrentCycle.Phase = CyclePhaseFailed
		c.CurrentCycle.CompletedAt = time.Now()
		c.CurrentCycle.Errors = append(c.CurrentCycle.Errors, err.Error())
		c.CycleHistory = append(c.CycleHistory, *c.CurrentCycle)
		_ = c.saveCycleHistory()
		return c.CurrentCycle, err
	}

	// Phase 3: Implementation
	if err = c.runImplementationPhase(); err != nil {
		c.CurrentCycle.Phase = CyclePhaseFailed
		c.CurrentCycle.CompletedAt = time.Now()
		c.CurrentCycle.Errors = append(c.CurrentCycle.Errors, err.Error())
		c.CycleHistory = append(c.CycleHistory, *c.CurrentCycle)
		_ = c.saveCycleHistory()
		return c.CurrentCycle, err
	}

	// Phase 4: Validation
	if err = c.runValidationPhase(); err != nil {
		c.CurrentCycle.Phase = CyclePhaseFailed
		c.CurrentCycle.CompletedAt = time.Now()
		c.CurrentCycle.Errors = append(c.CurrentCycle.Errors, err.Error())
		c.CycleHistory = append(c.CycleHistory, *c.CurrentCycle)
		_ = c.saveCycleHistory()
		return c.CurrentCycle, err
	}

	// Phase 5: Deployment
	if err = c.runDeploymentPhase(); err != nil {
		c.CurrentCycle.Phase = CyclePhaseFailed
		c.CurrentCycle.CompletedAt = time.Now()
		c.CurrentCycle.Errors = append(c.CurrentCycle.Errors, err.Error())
		c.CycleHistory = append(c.CycleHistory, *c.CurrentCycle)
		_ = c.saveCycleHistory()
		return c.CurrentCycle, err
	}

	// Mark as completed
	c.CurrentCycle.Phase = CyclePhaseCompleted
	c.CurrentCycle.CompletedAt = time.Now()

	// Calculate success rate
	if c.CurrentCycle.ImprovementsPlanned > 0 {
		c.CurrentCycle.SuccessRate = float64(c.CurrentCycle.ImprovementsValidated) / float64(c.CurrentCycle.ImprovementsPlanned)
	}

	c.CycleHistory = append(c.CycleHistory, *c.CurrentCycle)
	_ = c.saveCycleHistory()

	return c.CurrentCycle, nil
}

// runAnalysisPhase runs the analysis phase
func (c *ImprovementCycle) runAnalysisPhase() error {
	c.CurrentCycle.Phase = CyclePhaseAnalysis

	// Collect performance data
	performanceData, err := c.collectPerformanceData()
	if err != nil {
		return err
	}

	// Identify patterns
	patterns, err := c.identifyPatterns(performanceData)
	if err != nil {
		return err
	}

	c.CurrentCycle.PatternsIdentified = len(patterns)

	// Check success criteria
	if len(patterns) < c.phaseCriteria[CyclePhaseAnalysis].MinPatterns {
		return fmt.Errorf("insufficient patterns identified: %d", len(patterns))
	}

	return nil
}

// runPlanningPhase runs the planning phase
func (c *ImprovementCycle) runPlanningPhase() error {
	c.CurrentCycle.Phase = CyclePhasePlanning

	// Generate improvement plan
	plan, err := c.generateImprovementPlan()
	if err != nil {
		return err
	}

	c.CurrentCycle.ImprovementsPlanned = len(plan.Improvements)

	return nil
}

// runImplementationPhase runs the implementation phase
func (c *ImprovementCycle) runImplementationPhase() error {
	c.CurrentCycle.Phase = CyclePhaseImplementation

	// Implement improvements
	implemented, err := c.implementImprovements()
	if err != nil {
		return err
	}

	c.CurrentCycle.ImprovementsImplemented = implemented

	return nil
}

// runValidationPhase runs the validation phase
func (c *ImprovementCycle) runValidationPhase() error {
	c.CurrentCycle.Phase = CyclePhaseValidation

	// Validate improvements
	validated, err := c.validateImprovements()
	if err != nil {
		return err
	}

	c.CurrentCycle.ImprovementsValidated = validated

	return nil
}

// runDeploymentPhase runs the deployment phase
func (c *ImprovementCycle) runDeploymentPhase() error {
	c.CurrentCycle.Phase = CyclePhaseDeployment

	// Deploy improvements
	return c.deployImprovements()
}

// collectPerformanceData collects performance data from all agents
func (c *ImprovementCycle) collectPerformanceData() (map[string]interface{}, error) {
	data := map[string]interface{}{
		"timestamp":      time.Now().Format(time.RFC3339),
		"agents":         make(map[string]interface{}),
		"system_metrics": make(map[string]interface{}),
	}

	// Collect from each agent
	for agentType, agent := range c.Agents {
		performance := agent.(*BaseAgent).GetPerformanceSummary()
		data["agents"].(map[string]interface{})[agentType] = performance
	}

	return data, nil
}

// identifyPatterns identifies patterns in performance data
func (c *ImprovementCycle) identifyPatterns(performanceData map[string]interface{}) ([]map[string]interface{}, error) {
	// Pattern identification will be implemented with LLM service
	return []map[string]interface{}{}, nil
}

// generateImprovementPlan generates improvement plan
func (c *ImprovementCycle) generateImprovementPlan() (*ImprovementPlan, error) {
	return &ImprovementPlan{
		PlanID:    fmt.Sprintf("plan_%s", time.Now().Format("20060102_150405")),
		CreatedAt: time.Now(),
	}, nil
}

// implementImprovements implements improvements from the plan
func (c *ImprovementCycle) implementImprovements() (int, error) {
	// Implementation logic will be added
	return 0, nil
}

// validateImprovements validates implemented improvements
func (c *ImprovementCycle) validateImprovements() (int, error) {
	// Validation logic will be added
	return 0, nil
}

// deployImprovements deploys validated improvements
func (c *ImprovementCycle) deployImprovements() error {
	// Deployment logic will be added
	return nil
}

// saveCycleHistory saves cycle history to disk
func (c *ImprovementCycle) saveCycleHistory() error {
	historyPath := filepath.Join("agents", "memory", "cycle_history.json")
	if err := os.MkdirAll(filepath.Dir(historyPath), 0755); err != nil {
		return err
	}

	data, err := json.MarshalIndent(c.CycleHistory, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(historyPath, data, 0644)
}

// GetStatus returns current improvement cycle status
func (c *ImprovementCycle) GetStatus() map[string]interface{} {
	c.mu.RLock()
	defer c.mu.RUnlock()

	status := map[string]interface{}{
		"running":            c.running,
		"total_cycles":       len(c.CycleHistory),
		"agents_initialized": len(c.Agents),
	}

	if c.CurrentCycle != nil {
		status["current_cycle"] = c.CurrentCycle
	}

	successful := 0
	for _, cycle := range c.CycleHistory {
		if cycle.Phase == CyclePhaseCompleted {
			successful++
		}
	}
	status["successful_cycles"] = successful

	return status
}

// Shutdown shuts down the improvement cycle
func (c *ImprovementCycle) Shutdown() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Shutdown all agents
	for _, agent := range c.Agents {
		if baseAgent, ok := agent.(*BaseAgent); ok {
			_ = baseAgent.Shutdown()
		}
	}

	c.running = false
	return nil
}